### PR TITLE
hubble: add support for LostEvent from Hubble's ring buffer

### PR DESCRIPTION
**Pull request by rolinh**
This PR adds support to report events that are lost in Hubble's ring buffer. This may happen on invalid read in the ring buffer, typically when an entry has already been overwritten by the writer before the reader had a chance to read it.

In the process, I also changed the ring reader such that the buffer that is used in follow-mode is allocated by the ring reader instead of the `readFrom` method from the ring buffer. This should help write meaningful ring buffer benchmarks (cc @twpayne).

Please, read commit by commit for more details.

```release-note
Report events that are lost in Hubble's ring buffer.
```

### CODE REVIEW

File: pkg/hubble/container/ring_reader.go [hunk 1]
```diff
@@ -77,26 +85,42 @@ func (r *RingReader) Next() (*v1.Event, error) {
...
 
 	select {
-	case e, ok := <-r.c:
-		if !ok {
-			// channel can only be closed by readFrom if ctx is cancelled
-			r.c = nil
-			r.ctx = nil
-			return nil
+	case e := <-r.followChan:
+		if e != nil {
```
Comment (gandro) [inline pkg/hubble/container/ring_reader.go]: Why is this needed? `e == nil` should only ever happen if the channel is closed, in which case I prefer the check with `ok` instead of `nil`.

File: pkg/hubble/container/ring.go [hunk 2]
```diff
@@ -369,7 +373,26 @@ func (r *Ring) readFrom(ctx context.Context, read uint64, ch chan<- *v1.Event) {
...
+		default:
+			// The writer overwrote the entry before we had time to read it.
+			// Send a ListEvent to notify the read-miss.
+			now := time.Now().UTC()
+			select {
+			case ch <- &v1.Event{
+				Timestamp: &timestamp.Timestamp{
+					Seconds: int64(now.Unix()),
+					Nanos:   int32(now.Nanosecond()),
+				},
```
Comment (gandro) [inline pkg/hubble/container/ring.go:385]: I believe this breaks the "ring buffer timestamps are supposed to be monotonic" assumption. It will cause the ring reader to stop immediately if `until` is set here:

https://github.com/cilium/cilium/blob/8030a81dbbcc189814163742fdb32af740c2cb80/pkg/hubble/observer/local_observer.go#L385-L393

**Review by gandro — CHANGES_REQUESTED** [review]
I did not check the tests in detail yet, but I added two questions that stood out to me.

File: pkg/hubble/container/ring_reader.go (hunk 1 above)
Comment (rolinh) [reply pkg/hubble/container/ring_reader.go]: Mmh, indeed, I can't quite remember why this change (note that the channel is now closed by the ring reader). I reverted to checking if the channel is closed.

File: pkg/hubble/container/ring.go (hunk 2 above)
Comment (rolinh) [reply pkg/hubble/container/ring.go:385]: Very nice catch! What would you propose we do instead?

Comment (rolinh) [reply pkg/hubble/container/ring.go:385]: I gave this more thoughts. I think that lost events need to be treated as a special case, similarly to node status events, as users will never explicitly request them. I think that we should simply send them over whenever they're encountered and not apply any filters on it. This would be more inline with what should be expected and solve the timestamp issue.
I'll update the PR accordingly.

Comment (rolinh) [reply pkg/hubble/container/ring.go:385]: @gandro I updated the PR, PTAL.

File: pkg/hubble/container/ring.go [hunk 3]
```diff
@@ -278,108 +282,117 @@ func (r *Ring) read(read uint64) (*v1.Event, error) {
...
+				Event: &flowpb.LostEvent{
+					Source:        flowpb.LostEventSource_HUBBLE_RING_BUFFER,
+					NumEventsLost: 1,
```
Comment (gandro) [inline pkg/hubble/container/ring.go:388]: nit: Ideally, we would coalesce multiple LostEvents to avoid spamming the consumer if it falls behind a lot. We are e.g. doing something similar in
https://github.com/cilium/cilium/blob/641c0f9b3072a014c9541ecaa4e00a2b24c98d97/pkg/hubble/monitor/consumer.go#L85

But considering that it would make this complex code even more complex and that we are probably sunsetting this implementation soon, I'm fine to leave it as is.

File: pkg/hubble/container/ring_reader.go [hunk 4]
```diff
@@ -77,26 +85,44 @@ func (r *RingReader) Next() (*v1.Event, error) {
 func (r *RingReader) NextFollow(ctx context.Context) *v1.Event {
 	// if the context changed between invocations, we also have to restart
 	// readFrom, as the old readFrom instance will be using the old context.
-	if r.c == nil || r.ctx != ctx {
-		r.c = r.ring.readFrom(ctx, r.idx)
+	if r.ctx != ctx {
+		if r.followChan == nil {
+			r.followChan = make(chan *v1.Event, r.followChanLen)
```
Comment (gandro) [inline pkg/hubble/container/ring_reader.go:90]: I find it a bit unergonomic that in order to stop the ring reader properly, you always have to both cancel the context and close the ring reader. I think if you switch the context between calls to NextFollow, the old channel can actually never be closed. Is that an issue?

I have not thought about it too hard, but it be possible to close `r.followChan` once the context is cancelled like in the previous version. We still could have a `Close()` method used for testing.

**Review by gandro — COMMENTED** [review]
Overall looks good to me! One question wrt the new `Close` API

File: pkg/hubble/container/ring.go (hunk 3 above)
Comment (rolinh) [reply pkg/hubble/container/ring.go:388]: Such events could be coalesced in Relay as well to mitigate this issue. Note that I don't expect so many of these events in practice.

File: pkg/hubble/container/ring_reader.go (hunk 4 above)
Comment (rolinh) [reply pkg/hubble/container/ring_reader.go:90]: That's a good feedback. I updated the code to make `Close()` optional as it's mostly useful for testing only.

File: pkg/hubble/container/ring_reader.go [hunk 5]
```diff
@@ -77,26 +85,50 @@ func (r *RingReader) Next() (*v1.Event, error) {
...
+
+// Close waits for any method to return and closes the RingReader.
+// Callers are not required to call Close on a RingReader but it may be useful
+// for specific situations such as testing.
+func (r *RingReader) Close() error {
+	r.wg.Wait()
+	if r.followChan != nil {
+		close(r.followChan)
+		r.followChan = nil
+	}
```
Comment (gandro) [inline pkg/hubble/container/ring_reader.go]: I think this is now not needed anymore. With the closure of the channel on line 94, there is no case where `r.wg.Wait()` unblocks without the channel already being closed, right?

File: pkg/hubble/container/ring_reader.go (hunk 4 above)
Comment (gandro) [reply pkg/hubble/container/ring_reader.go:90]: Awesome, one minor follow-up below

**Review by gandro — APPROVED** [review]

**Comment by Rolinh** [comment]
test-me-please

**Comment by Rolinh** [comment]
retest-net-next

**Review by aanm — APPROVED** [review]

**Review by twpayne — APPROVED** [review]
